### PR TITLE
Master accounts budget new api avo

### PR DESCRIPTION
**Pull request by avo-odoo**
task:-https://www.odoo.com/web#id=12160&view_type=form&model=project.task&menu_id=3942&action=327
pad:-https://pad.odoo.com/p/openerp-project.task-9DNPPA8H5M

### CODE REVIEW

File: addons/account_budget/__init__.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # -*- coding: utf-8 -*-
 ##############################################################################
 #
-#    OpenERP, Open Source Management Solution
+#    Odoo, Open Source Management Solution
 #    Copyright (C) 2004-2010 Tiny SPRL (<http://tiny.be>).
```
Comment (mdi-odoo) [inline addons/account_budget/__init__.py]: Tiny SPRL (<http://tiny.be>) ???

File: addons/account_budget/account_budget.py [hunk 2]
```diff
@@ -1,210 +1,162 @@
...
-    }
-    _defaults = {
-        'company_id': lambda self, cr, uid, c: self.pool.get('res.company')._company_default_get(cr, uid, 'account.budget.post', context=c)
-    }
+
+    code = fields.Char(required=True)
+    name = fields.Char(required=True)
+    account_ids = fields.Many2many('account.account', 'account_budget_rel', 'budget_id', 'account_id', string='Accounts', domain=[('deprecated', '=', False)])
+    crossovered_budget_line = fields.One2many('crossovered.budget.lines', 'general_budget_id', string='Budget Lines')
+    company_id = fields.Many2one('res.company', string='Company', required=True, default=lambda self: self.env.user.company_id)
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: Use:
default=lambda self: self.env['res.company']._company_default_get('account.budget.post')
for default value.

File: addons/account_budget/account_budget.py [hunk 3]
```diff
@@ -1,210 +1,162 @@
...
-        self.write(cr, uid, ids, {
+    name = fields.Char(required=True, states={'done':[('readonly',True)]})
+    code = fields.Char(required=True, states={'done':[('readonly',True)]})
+    creating_user_id = fields.Many2one('res.users', string='Responsible User', default=lambda self: self.env.uid)
+    validating_user_id = fields.Many2one('res.users', string='Validate User', readonly=True)
+    date_from = fields.Date(string='Start Date', required=True, states={'done':[('readonly',True)]})
+    date_to = fields.Date(string='End Date', required=True, states={'done':[('readonly',True)]})
+    state = fields.Selection([('draft','Draft'),('cancel', 'Cancelled'),('confirm','Confirmed'),('validate','Validated'),('done','Done')], index=True, required=True, readonly=True, copy=False, default='draft')
+    crossovered_budget_line = fields.One2many('crossovered.budget.lines', 'crossovered_budget_id', string='Budget Lines', states={'done':[('readonly',True)]}, copy=True)
+    company_id = fields.Many2one('res.company', string='Company', required=True, default=lambda self: self.env.user.company_id)
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: Wrong value for default !!!

File: addons/account_budget/account_budget.py [hunk 4]
```diff
@@ -1,210 +1,162 @@
...
-    _defaults = {
-        'state': 'draft',
-        'creating_user_id': lambda self, cr, uid, context: uid,
-        'company_id': lambda self, cr, uid, c: self.pool.get('res.company')._company_default_get(cr, uid, 'account.budget.post', context=c)
-    }
-
-    def budget_confirm(self, cr, uid, ids, *args):
-        self.write(cr, uid, ids, {
+    name = fields.Char(required=True, states={'done':[('readonly',True)]})
+    code = fields.Char(required=True, states={'done':[('readonly',True)]})
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: Why you removed the size ???

File: addons/account_budget/account_budget.py [hunk 5]
```diff
@@ -1,210 +1,162 @@
...
 
 
-class crossovered_budget_lines(osv.osv):
+class CrossoveredBudgetLines(models.Model):
 
-    def _prac_amt(self, cr, uid, ids, context=None):
-        res = {}
+    @api.multi
+    def _prac_amt(self):
+        self.ensure_one()
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: self.ensure_one() ???

File: addons/account_budget/account_budget.py [hunk 6]
```diff
@@ -1,210 +1,162 @@
...
-
-            if line.paid_date:
-                if strToDate(line.date_to) <= strToDate(line.paid_date):
-                    theo_amt = 0.00
-                else:
-                    theo_amt = line.planned_amount
+        acc_ids = [x.id for x in self.general_budget_id.account_ids]
+        if not acc_ids:
+            raise UserError(_("The Budget '%s' has no accounts!") % ustr(self.general_budget_id.name))
+        if self.analytic_account_id.id:
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: It seems that complete re-factoring of method is needed !!!

File: addons/account_budget/account_budget.py [hunk 7]
```diff
@@ -1,210 +1,162 @@
...
+        return result
+
+    @api.multi
+    def _prac(self):
+        for line in self:
+            line.practical_amount = line._prac_amt()
+
+    @api.multi
+    def _theo_amt(self):
+        self.ensure_one()
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: self.ensure_one() ???

File: addons/account_budget/account_budget.py [hunk 8]
```diff
@@ -1,210 +1,162 @@
...
+            elapsed_timedelta = today - (strToDatetime(self.date_from))
+
+            if elapsed_timedelta.days < 0:
+                # If the budget line has not started yet, theoretical amount should be zero
+                theo_amt = 0.00
+            elif line_timedelta.days > 0 and today < strToDatetime(self.date_to):
+                # If today is between the budget line date_from and date_to
+                theo_amt = (elapsed_timedelta.total_seconds() / line_timedelta.total_seconds()) * self.planned_amount
+            else:
+                theo_amt = self.planned_amount
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: It seems that complete re-factoring of method is needed !!!

File: addons/account_budget/account_budget.py [hunk 9]
```diff
@@ -1,210 +1,162 @@
...
+        if result is None: result = 0.00
+        return result
+
+    @api.multi
+    def _prac(self):
+        for line in self:
+            line.practical_amount = line._prac_amt()
+
+    @api.multi
+    def _theo_amt(self):
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: Specify the correct field dependencies of a "compute" method !!!

File: addons/account_budget/account_budget.py [hunk 10]
```diff
@@ -1,210 +1,162 @@
...
+            elif line_timedelta.days > 0 and today < strToDatetime(self.date_to):
+                # If today is between the budget line date_from and date_to
+                theo_amt = (elapsed_timedelta.total_seconds() / line_timedelta.total_seconds()) * self.planned_amount
+            else:
+                theo_amt = self.planned_amount
+
+        return theo_amt
+
+    @api.multi
+    def _theo(self):
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: Specify the correct field dependencies of a "compute" method !!!

File: addons/account_budget/account_budget.py [hunk 11]
```diff
@@ -1,210 +1,162 @@
...
+        for line in self:
+            line.theoritical_amount = line._theo_amt()
+
+    @api.multi
+    def _perc(self):
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: Specify the correct field dependencies of a "compute" method !!!

File: addons/account_budget/account_budget.py [hunk 12]
```diff
@@ -1,210 +1,162 @@
...
-        'percentage':fields.function(_perc, string='Percentage', type='float'),
-        'company_id': fields.related('crossovered_budget_id', 'company_id', type='many2one', relation='res.company', string='Company', store=True, readonly=True)
-    }
-
-
-class account_analytic_account(osv.osv):
+
+    crossovered_budget_id = fields.Many2one('crossovered.budget', string='Budget', ondelete='cascade', index=True, required=True)
+    analytic_account_id = fields.Many2one('account.analytic.account', string='Analytic Account')
+    general_budget_id = fields.Many2one('account.budget.post', string='Budgetary Position',required=True)
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: Follow PEP 8 standard - Use space after comma !!!

File: addons/account_budget/wizard/account_budget_analytic.py [hunk 13]
```diff
@@ -1,49 +1,24 @@
...
-        'date_from': lambda *a: time.strftime('%Y-01-01'),
-        'date_to': lambda *a: time.strftime('%Y-%m-%d'),
-    }
 
-    def check_report(self, cr, uid, ids, context=None):
-        if context is None:
-            context = {}
-        data = self.read(cr, uid, ids, context=context)[0]
+    date_from = fields.Date(string='Start of period', required=True, default=lambda *a: time.strftime('%Y-01-01'))
+    date_to = fields.Date(string='End of period', required=True, default=lambda *a: time.strftime('%Y-%m-%d'))
```
Comment (mdi-odoo) [inline addons/account_budget/wizard/account_budget_analytic.py]: Use default=fields.Date.context_today instead !!!

File: addons/account_budget/wizard/account_budget_crossovered_report.py [hunk 14]
```diff
@@ -1,50 +1,25 @@
...
-        'date_from': lambda *a: time.strftime('%Y-01-01'),
-        'date_to': lambda *a: time.strftime('%Y-%m-%d'),
-    }
 
-    def check_report(self, cr, uid, ids, context=None):
-        if context is None:
-            context = {}
-        data = self.read(cr, uid, ids, context=context)[0]
+    date_from = fields.Date(string='Start of period', required=True, default=lambda *a: time.strftime('%Y-01-01'))
+    date_to = fields.Date(string='End of period', required=True, default=lambda *a: time.strftime('%Y-%m-%d'))
```
Comment (mdi-odoo) [inline addons/account_budget/wizard/account_budget_crossovered_report.py]: Use default=fields.Date.context_today instead !!!

File: addons/account_budget/__init__.py (hunk 1 above)
Comment (avo-odoo) [inline addons/account_budget/__init__.py]: Done.

File: addons/account_budget/account_budget.py (hunk 2 above)
Comment (avo-odoo) [inline addons/account_budget/account_budget.py]: Done.

File: addons/account_budget/account_budget.py (hunk 3 above)
Comment (avo-odoo) [inline addons/account_budget/account_budget.py]: Done.

File: addons/account_budget/account_budget.py (hunk 4 above)
Comment (avo-odoo) [inline addons/account_budget/account_budget.py]: add back size attribute

File: addons/account_budget/account_budget.py (hunk 5 above)
Comment (avo-odoo) [inline addons/account_budget/account_budget.py]: removed it.

File: addons/account_budget/account_budget.py (hunk 6 above)
Comment (avo-odoo) [inline addons/account_budget/account_budget.py]: Done.

File: addons/account_budget/account_budget.py (hunk 7 above)
Comment (avo-odoo) [inline addons/account_budget/account_budget.py]: Done.removed it

File: addons/account_budget/account_budget.py (hunk 8 above)
Comment (avo-odoo) [inline addons/account_budget/account_budget.py]: Done.

File: addons/account_budget/account_budget.py (hunk 9 above)
Comment (avo-odoo) [inline addons/account_budget/account_budget.py]: Done.

File: addons/account_budget/account_budget.py (hunk 10 above)
Comment (avo-odoo) [inline addons/account_budget/account_budget.py]: Done.

File: addons/account_budget/account_budget.py (hunk 11 above)
Comment (avo-odoo) [inline addons/account_budget/account_budget.py]: Done.

File: addons/account_budget/account_budget.py (hunk 12 above)
Comment (avo-odoo) [inline addons/account_budget/account_budget.py]: Done.

File: addons/account_budget/wizard/account_budget_analytic.py (hunk 13 above)
Comment (avo-odoo) [inline addons/account_budget/wizard/account_budget_analytic.py]: Done.

File: addons/account_budget/wizard/account_budget_crossovered_report.py (hunk 14 above)
Comment (avo-odoo) [inline addons/account_budget/wizard/account_budget_crossovered_report.py]: Done.

File: addons/account_budget/account_budget.py [hunk 15]
```diff
@@ -1,154 +1,122 @@
...
-        'name': fields.char('Name', required=True),
-        'account_ids': fields.many2many('account.account', 'account_budget_rel', 'budget_id', 'account_id', 'Accounts', domain=[('deprecated', '=', False)]),
-        'crossovered_budget_line': fields.one2many('crossovered.budget.lines', 'general_budget_id', 'Budget Lines'),
-        'company_id': fields.many2one('res.company', 'Company', required=True),
-    }
-    _defaults = {
-        'company_id': lambda self, cr, uid, c: self.pool.get('res.company')._company_default_get(cr, uid, 'account.budget.post', context=c)
-    }
+
+    code = fields.Char(required=True)
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: Why you removed size from 'code' field ???

File: addons/account_budget/account_budget.py [hunk 16]
```diff
@@ -1,154 +1,122 @@
...
-    _defaults = {
-        'state': 'draft',
-        'creating_user_id': lambda self, cr, uid, context: uid,
-        'company_id': lambda self, cr, uid, c: self.pool.get('res.company')._company_default_get(cr, uid, 'account.budget.post', context=c)
-    }
-
-    def budget_confirm(self, cr, uid, ids, *args):
-        self.write(cr, uid, ids, {
+    name = fields.Char(required=True, states={'done':[('readonly',True)]})
+    code = fields.Char(required=True, size=16, states={'done':[('readonly',True)]})
```
Comment (mdi-odoo) [inline addons/account_budget/account_budget.py]: Follow PEP 8 standard - Use space after comma !!!

File: addons/account_budget/report/analytic_account_budget_report.py [hunk 17]
```diff
@@ -1,32 +1,13 @@
...
-##############################################################################
 
 import time
-from openerp.osv import osv
+from openerp import models
 from openerp.report import report_sxw
 
 
-class analytic_account_budget_report(report_sxw.rml_parse):
+class AnalyticAccountBudgetReport(report_sxw.rml_parse):
```
Comment (mdi-odoo) [inline addons/account_budget/report/analytic_account_budget_report.py]: Need to convert the report to new style .

File: addons/account_budget/report/budget_report.py [hunk 18]
```diff
@@ -1,33 +1,14 @@
...
 
 import time
-from openerp.osv import osv
+from openerp import models
 from openerp.report import report_sxw
 
 tot = {}
 
-class budget_report(report_sxw.rml_parse):
+class BudgetReport(report_sxw.rml_parse):
```
Comment (mdi-odoo) [inline addons/account_budget/report/budget_report.py]: Convert the report to new style .

**Pull request closed without merging**